lib/std/simd.zig

```diff
@@ -0,0 +1,369 @@
+    try std.testing.expectEqual([4]u32{ 20, 30, 40, 10 }, rotateElementsDown(base, 1));
+    try std.testing.expectEqual([4]u32{ 40, 10, 20, 30 }, rotateElementsUp(base, 1));
+    try std.testing.expectEqual([4]u32{ 40, 30, 20, 10 }, reverse(base));
+}
+
+pub fn firstTrue(vec: anytype) ?VectorIndex(@TypeOf(vec)) {
+    const len = vectorLength(@TypeOf(vec));
+    const IndexInt = VectorIndex(@TypeOf(vec));
+
+    if (!@reduce(.Or, vec)) {
+        return null;
+    }
+    const indexes = @select(IndexInt, vec, countUp(IndexInt, len), @splat(len, ~@as(IndexInt, 0)));
+    return @reduce(.Min, indexes);
+}
+
+pub fn lastTrue(vec: anytype) ?VectorIndex(@TypeOf(vec)) {
+    const len = vectorLength(@TypeOf(vec));
+    const IndexInt = VectorIndex(@TypeOf(vec));
+
+    if (!@reduce(.Or, vec)) {
+        return null;
+    }
+    const indexes = @select(IndexInt, vec, countUp(IndexInt, len), @splat(len, @as(IndexInt, 0)));
+    return @reduce(.Max, indexes);
+}
+
+pub fn countTrues(vec: anytype) VectorCount(@TypeOf(vec)) {
+    const len = vectorLength(@TypeOf(vec));
+    const CountIntType = VectorCount(@TypeOf(vec));
+
+    const one_if_true = @select(CountIntType, vec, @splat(len, @as(CountIntType, 1)), @splat(len, @as(CountIntType, 0)));
+    return @reduce(.Add, one_if_true);
+}
+
+pub fn firstIndexOfValue(vec: anytype, value: std.meta.Child(@TypeOf(vec))) ?VectorIndex(@TypeOf(vec)) {
+    const len = vectorLength(@TypeOf(vec));
+
+    return firstTrue(vec == @splat(len, value));
+}
+
+pub fn lastIndexOfValue(vec: anytype, value: std.meta.Child(@TypeOf(vec))) ?VectorIndex(@TypeOf(vec)) {
+    const len = vectorLength(@TypeOf(vec));
+
+    return lastTrue(vec == @splat(len, value));
+}
+
+pub fn countElementsWithValue(vec: anytype, value: std.meta.Child(@TypeOf(vec))) VectorCount(@TypeOf(vec)) {
+    const len = vectorLength(@TypeOf(vec));
+
+    return countTrues(vec == @splat(len, value));
+}
+
+test "vector searching" {
+    const base = Vector(8, u32){ 6, 4, 7, 4, 4, 2, 3, 7 };
+
+    try std.testing.expectEqual(@as(?u3, 1), firstIndexOfValue(base, 4));
+    try std.testing.expectEqual(@as(?u3, 4), lastIndexOfValue(base, 4));
+    try std.testing.expectEqual(@as(?u3, null), lastIndexOfValue(base, 99));
+    try std.testing.expectEqual(@as(u4, 3), countElementsWithValue(base, 4));
+}
+
+/// Same as accum, but with a user-provided function that mathematically must be associative.
+/// The function must also do nothing when called with identity as either argument.
+pub fn accumWithFunc(comptime hop: isize, vec: anytype, comptime func: fn (@TypeOf(vec), @TypeOf(vec)) @TypeOf(vec), comptime identity: std.meta.Child(@TypeOf(vec))) @TypeOf(vec) {
+    // I haven't debugged this, but it might be a cousin of sorts to what's going on with braid.
+    comptime if (builtin.cpu.arch.isMIPS() and !builtin.is_test) @compileError("TODO: Find out why accum doesn't work on MIPS");
```

Comment:
Does it only break without test mode? You did not mention this in the associated text.

Reply:
This is just me struggling to stop the compile error from triggering under the wrong circumstances. Not having the `comptime` keyword caused the compile error to trigger on every platform. Admittedly, it was a bit rushed, so I'll try removing the `!builtin.is_test` and see what happens once I get home.